man/redirect.8
.\"                                                              -*- nroff -*-
.\" The Merecat web server stems from both sthttpd and thttpd, both of
.\" which are free software under the 2-clause simplified BSD license.
.\"
.\" Copyright (c) 1995-2015  Jef Poskanzer <jef@mail.acme.com>
.\" All rights reserved.
.\"
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\"
.\" THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
.\" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
.\" IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
.\" ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNERS OR CONTRIBUTORS BE
.\" LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
.\" CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
.\" SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
.\" INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
.\" CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
.\" ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
.\" THE POSSIBILITY OF SUCH DAMAGE.
.Dd Aug 3, 2019
.Dt REDIRECT 8 SMM
.Os "merecat (2.32)"
.Sh NAME
.Nm redirect
.Nd Simple redirection CGI program
.Sh SYNOPSIS
.Nm
.Sh DESCRIPTION
Three steps to set up a redirection:
.Bl -enum
.It
Make sure your web server is set up to allow CGI programs.
.It
Make a symbolic link from the file or directory you want to redirect,
pointing at this program in the CGI bin directory.
.It
Add an entry to the file ".redirects" in the directory where your http
server runs CGI programs.  For most servers, this is the directory where
the given CGI program lives.
.El
.Sh SYNTAX
The format of the .redirects file is a bunch of lines with a filename,
whitespace, and the new URL.  For example:
.Bd -unfilled -offset left

    /test/oldfile.html    http://www.acme.com/test/newfile.html

.Ed
The easiest way to figure out precisely what filename to put
into .redirects is to set up the symlink and then click on it.  You'll
get back a "404 Not Found" page which includes the filename as received
by the redirect program, and that's what you want to use.
.Pp
You can also add a wildcard specification to redirect whole groups of
files.  For example:
.Bd -unfilled -offset left

    /wildtest/*          http://www.acme.com/test-

.Ed
will cause an access to the /wildtest/somefile.html to be redirected to
http://www.acme.com/test-somefile.html. (Note that the asterisk need not
be preceded by a slash.)
.Pp
.Em Note:
.Nm
is designed for use with
.Xr thttpd 8
and
.xr merecat 8,
but the latter has its own built-in redirect support today which is
probably what you want.  Using
.Nm
with other web servers may require some hacking.  A possible gotcha is
with the symbolic link from the old file pointing at this script -
servers other than thttpd may not allow that link to be run as a CGI
program, because they don't check the link to see that it points into
the allowed CGI directory.
.Sh SEE ALSO
.Xr thttpd 8 ,
.Xr merecat 8 ,
.Xr merecat.conf 5
.Sh AUTHORS
.An Jef Poskanzer Aq jef@mail.acme.com
.Sh BUGS
It would be really cool to have this program look for the .redirects
file in the same directory as the file being redirected, instead of in
the binaries directory.  Unfortunately, this appears to be impossible
with the information CGI gives, plus the non-standardized but widespread
practice of running CGI programs in the directory where the binary
lives.  Perhaps CGI 1.2 will address this.
.Pp
The wildcard mechanism is very primitive.  In particular, any characters
that follow the asterisk are blithely ignored.


ChangeLog.md
Change Log
==========

All relevant changes are documented in this file.


[v2.32][] - 2019-07-XX
----------------------

Notable new features: multiple server support from one process, HTTPS,
HTTP/1.1 keep-alive, and built-in gzip deflate compression using zlib.

### Changes
- Add support for HTTPS, works with certificates from Let's Encrypt
- Add support for multiple servers, listen to different ports
- Add support for built-in HTTP redirect, e.g. from HTTP to HTTPS
- Add gzip deflate compression when built with zlib, also compress
  HEAD as well as GET requests
- Add true `Connection: keep-alive` support
- Add missing `Vary: Accept-Encoding` header
- CGI: Allow handling other HTTP methods besides GET/HEAD/POST, from
  thttpd v2.29, change by Jef Poskanzer
- CGI: Allow `:PORT` in `HTTP_POST`, like Apache
- CGI: Allow trailing slash in `PATH_INFO`, like Apache
- CGI: Change default `CGI_PATTERN` from disabled to `**.cgi|/cgi-bin/*`
- CGI: Add support for looking for an `index.cgi` index file
- CGI: Add several missing standard CGI/1.1 environment variables, see
  the file docs/cgi.txt for details
- PHP:
  - Add support for `php-cgi` and `index.php` index file
  - Add support for PHP pattern matching, run php-cgi if `**.php`
- Server-Side Includes (SSI):
  - Add support for SSI pattern matching, run cgi-bin/ssi if `**.shtml`
  - Add support for silencing default SSI `errmsg`
  - Add support for looking for `index.shtml` index file
- Dot files are no longer shown in dir listings, use the `merecat.conf`
  setting `list-dotfiles = true` to enable
- Server stats are no longer periodically sent to syslog, re-enable in
  `merecat.conf` if you need the `STATS_TIME` feature
- Apply Debian thttpd `SIGBUS` patch for reading from NFS
- Add `-I IDENT` command line option to override program identity.
  This change makes it possible to change syslog, PID file name, *and*
  `.conf` file name.  Useful when running multiple instances of Merecat
- Add `--enable-msie-padding` to `configure` script
- Add `.htaccess` support, limited to IPv4.  Feature by Felix J. Ogris
- Allow `.htpasswd` file to be symlinked
- DOC: How to use `.htpasswd` and virtual hosts
- DOC: Added section on how to optimize performance
- Update MIME types, e.g. Ogg video, 7zip, svg
- Add Dockerfile for ease of deployment in limited setups
- Add cute cat default favicon
- Built-in icons for FTP dir listings; folder, file, etc.
- Refactor, deprecated POSIX API's, e.g. `bzero() --> memset()`
- Enable `SO_REUSEPORT` if available, useful for load balancing

### Fixes
- Fix CVE-2017-17663, buffer overrun in htpasswd tool, from thttpd v2.28
- Fixes for non GNU C libraries like musl: `__progname`, `%m`, etc.
- Fix `X-Forwarded-For` when using IPv6, thanks to Steve Kemp!
- Debian packaging fixes
- Make sure both `.htpasswd` *and* `.htaccess` are declared forbidden
  files and not allowed to be downloaded or shown in directory listings
- Use `memmove()` instead of `strcpy()` for possibly overlapping regions
- Cleanup of default `merecat.conf`, default disabled options to their
  built-in default values
- Spelling fixes and major documentation cleanup


[v2.31][] - 2016-11-06
----------------------

The "it works now" release.

### Changes
- Sort directories first in dir listings
- Include systemd unit file
- Add `debian/` packaging, easy to rebuild and replace for others
- Add `--enable-public-html` to enable `~user/public_html` dirs
- Support for shared `WEBROOT/cgi-bin` as fallback for vhosts
- Update default landing page

### Fixes
- Add missing CSS and jpeg files to install
- Fix dependency tracking when reconfiguring
- Fix `.conf` file parser bugs reported by Gaetan Bisson
- Fix missing `HAVE_LIBCONFUSE` #define causing `.conf` file support to
  not be built, reported by Gaetan Bisson
- Fix malplaced call to `cfg_free()` in .conf file parser, reported by
  Gaetan Bisson
- Update man page and other documentation with missing quotes around CGI
  pattern, issue reported by Gaetan Bisson
- Fix syslog warning: bind 0.0.0.0: Address already in use


[v2.30][] - 2016-10-09
----------------------

Initial release.  Based on [sthttpd][] master, 2015-07-22.

[sthttpd]: https://github.com/blueness/sthttpd/


www/cgi-bin/redirect.c
/* redirect - simple redirection CGI program
**
** Copyright (C) 1995  Jef Poskanzer <jef@mail.acme.com>
** All rights reserved.
**
** Redistribution and use in source and binary forms, with or without
** modification, are permitted provided that the following conditions
** are met:
** 1. Redistributions of source code must retain the above copyright
**    notice, this list of conditions and the following disclaimer.
** 2. Redistributions in binary form must reproduce the above copyright
**    notice, this list of conditions and the following disclaimer in the
**    documentation and/or other materials provided with the distribution.
**
** THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
** AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
** IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
** ARE DISCLAIMED.  IN NO EVENT SHALL THE COPYRIGHT OWNERS OR CONTRIBUTORS BE
** LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR
** CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
** SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
** INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
** CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
** ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF
** THE POSSIBILITY OF SUCH DAMAGE.
*/

/* Three steps to set up a redirection:
**  1. Make sure your web server is set up to allow CGI programs.
**  2. Make a symbolic link from the file you want to redirect,
**     pointing at this program in the CGI bin directory.
**  3. Add an entry to the file ".redirects" in the directory where your
**     http server runs CGI programs.  For most servers, this is the
**     directory where the given CGI program lives.  The format of the
**     file is a bunch of lines with a filename, whitespace, and the new
**     URL.  For example:

/test/oldfile.html    http://www.acme.com/test/newfile.html

**     The easiest way to figure out precisely what filename to put into
**     .redirects is to set up the symlink and then click on it.  You'll get
**     back a "404 Not Found" page which includes the filename as received by
**     the redirect program, and that's what you want to use.
**
** Note: this is designed for thttpd (http://www.acme.com/software/thttpd/)
** and using it with other web servers may require some hacking.  A possible
** gotcha is with the symbolic link from the old file pointing at this
** script - servers other than thttpd may not allow that link to be run
** as a CGI program, because they don't check the link to see that it
** points into the allowed CGI directory.
**
** Note two: It would be really cool to have this program look for
** the .redirects file in the same directory as the file being redirected,
** instead of in the binaries directory.  Unfortunately, this appears
** to be impossible with the information CGI gives, plus the non-standardized
** but widespread practice of running CGI programs in the directory where
** the binary lives.  Perhaps CGI 1.2 will address this.
*/

#include <config.h>

/* System headers */
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <sys/types.h>

/* Local headers */
#include "merecat.h"

static void internal_error(char *reason)
{
	char *title = "500 Internal Error";

	printf("\
Status: %s\n\
Content-type: text/html\n\
\n\
<HTML><HEAD><TITLE>%s</TITLE></HEAD>\n\
<BODY><H2>%s</H2>\n\
Something unusual went wrong during a redirection request:\n\
<BLOCKQUOTE>\n\
%s\n\
</BLOCKQUOTE>\n\
</BODY></HTML>\n", title, title, title, reason);
}


static void not_found(char *script_name)
{
	char *title = "404 Not Found";

	printf("\
Status: %s\n\
Content-type: text/html\n\
\n\
<HTML><HEAD><TITLE>%s</TITLE></HEAD>\n\
<BODY><H2>%s</H2>\n\
The requested filename, %s, is set up to be redirected to another URL;\n\
however, the new URL has not yet been specified.\n\
</BODY></HTML>\n", title, title, title, script_name);
}


static void moved(char *script_name, char *url)
{
	char *title = "Moved";

	printf("\
Location: %s\n\
Content-type: text/html\n\
\n\
<HTML><HEAD><TITLE>%s</TITLE></HEAD>\n\
<BODY><H2>%s</H2>\n\
The requested filename, %s, has moved to a new URL:\n\
<A HREF=\"%s\">%s</A>.\n\
</BODY></HTML>\n", url, title, title, script_name, url, url);
}


int main(int argc, char **argv)
{
	char *script_name;
	char *path_info;
	char *cp;
	FILE *fp;
	char *star;
	char buf[5000], file[5000], url[5000];

	/* Get the name that we were run as, which is the filename being
	 ** redirected.
	 */
	script_name = getenv("SCRIPT_NAME");
	if (script_name == (char *)0) {
		internal_error("Couldn't get SCRIPT_NAME environment variable.");
		exit(1);
	}

	/* Append the PATH_INFO, if any.  This allows redirection of whole
	 ** directories.
	 */
	path_info = getenv("PATH_INFO");
	if (path_info != (char *)0) {
		cp = (char *)malloc(strlen(script_name) + strlen(path_info) + 1);
		if (cp == (char *)0) {
			internal_error("Out of memory.");
			exit(1);
		}
		sprintf(cp, "%s%s", script_name, path_info);
		script_name = cp;
	}

	/* Open the redirects file. */
	fp = fopen(".redirects", "r");
	if (fp == (FILE *)0) {
		internal_error("Couldn't open .redirects file.");
		exit(1);
	}

	/* Search the file for a matching entry. */
	while (fgets(buf, sizeof(buf), fp) != (char *)0) {
		/* Remove comments. */
		cp = strchr(buf, '#');
		if (cp != (char *)0)
			*cp = '\0';
		/* Skip leading whitespace. */
		cp = buf;
		cp += strspn(cp, " \t");
		/* Check for blank line. */
		if (*cp != '\0') {
			/* Parse line. */
			if (sscanf(cp, "%5000[^ \t\n] %5000[^ \t\n]", file, url) == 2) {
				/* Check for wildcard match. */
				star = strchr(file, '*');
				if (star != (char *)0) {
					/* Check for leading match. */
					if (strncmp(file, script_name, star - file) == 0) {
						/* Got it; put together the full name. */
						strcat(url, script_name + (star - file));
						/* XXX Whack the script_name, too? */
						moved(script_name, url);
						exit(0);
					}
				}
				/* Check for exact match. */
				if (strcmp(file, script_name) == 0) {
					/* Got it. */
					moved(script_name, url);
					exit(0);
				}
			}
		}
	}

	/* No match found. */
	not_found(script_name);
	exit(1);
}
